### PR TITLE
Perf: plonk verifier gadget

### PR DESCRIPTION
**Pull request by yelhousni**
# Description

*(first review #874 ~and https://github.com/Consensys/gnark-crypto/pull/471~)*

Some optimizations for the PLONK verifier gadget:
- [x] precomputed lines in pairing (#874 merged here)
- [x] `JointScalarMul` (`s1*p1+s2*p2` using Shamir's trick)
- [x] `scalarBitsMul` (scalar binary decomposition only once)
- [x] `MultiScalarMul` to use `JointScalarMul` and `scalarBitsMul` in the folding case
- [x] Revisit Karabina cyclotomic square (a variant that is more efficient to decompress in-circuit for some relevant sizes)
- [x] re-arrange addition chains to make optimal use of the new Karabina square
- [x] rewrite Hayashida et al. hard part for BLS12 (observe that `(x-1)^2` has a better addition chain than `x` or `x^2`)
- [x] addition vs. unified addition when relevant
- [x] nonzero scalar and point handling only when relevant
- [x] avoid emulated negations and handle trivial mul/add by 0/1
- [x] use of endomorphism for `ScalarMul` and `MultiScalarMul` (https://eprint.iacr.org/2019/1021, sec. 6.2)
- [ ] ~use `1/x` as a bijection in bw6 MiMC instead of `x^5`(needs https://github.com/Consensys/gnark-crypto/pull/471)~

TODO: 
- [ ] Augment the verifying key with some pre-computed data related to the fixed points:
    - [ ] binary decompositions for `MarshalG1`
    - [ ] doubles of points for `ScalarMul` and `MultiScalarMul`

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How has this been tested?

All current tests pass.

# How has this been benchmarked?

- 342,254 SCS for the bls12-377 in bw6-761 case.
- 44,095,764 SCS for the bw6-761 in bn254 case.

# Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I did not modify files generated from templates
- [x] `golangci-lint` does not output errors locally
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by ivokub — CHANGES_REQUESTED** [review]
Nice! All seems good. I already pushed some cosmetic changes. I think there are still a few issues:
* for two-chain we have wrapper for MillerLoop which goes from `[]G2Affine` argument to `[]*G2Affine`. But we compute the lines in the method which uses `[]G2Affine`. However, as the slice contains of values then inside a method we are working on a copy of `G2Affine` value and we are not updating the lazy lines of the input. I think we can either change `MillerLoop` to take `[]*G2Affine` so that can modify inline or precompute the lines in `Pairing` wrapper method `Pairing.MillerLoop` to already precompute. I like the first approach better because then we always will use lazy line computation.
* I would still like to get PLONK verification without commitment working. I'm actually thinking about adding new `algopt` option `algopts.ForceSafe` which performs safe arithmetic inside MSM. It is suboptimal (we could also only omit the selector we know is 0), but atleast would cover different cases. And I think we already know inside the circuit if there is commitment or not.
* there are a few unused functions. I didn't want to remove yet as maybe could be useful in the future - for example Pairing.generators() in packages or addStep() in some packages. If you say they are good to remove then I can.
* I'm not sure, but isn't there regression for ScalarMulBase in 2-chains? Previously we had computed powers of twos, but now we're using generic scalar multiplication (with GLV nontheless). Have you run any comparisons which is better for base scalar mul - generic with GLV or double-and-add with precomputed points.

**Review by ivokub — APPROVED** [review]
The PR is good to merge. We can address my other issues in other PR. I think this PR needs to be merged to unblock #960.

File: std/algebra/native/sw_bls12377/g1.go [hunk 1]
```diff
@@ -453,33 +511,158 @@ func (p *G1Affine) DoubleAndAdd(api frontend.API, p1, p2 *G1Affine) *G1Affine {
...
+	P.X = Acc.X
+	P.Y = Acc.Y
 
-	P.X = res.X
-	P.Y = res.Y
+	return P
+}
+
+// scalarBitsMul...
+func (P *G1Affine) scalarBitsMul(api frontend.API, Q G1Affine, s1bits, s2bits []frontend.Variable, opts ...algopts.AlgebraOption) *G1Affine {
```
[the remark this answers is not in the document]
Comment (yelhousni) [reply std/algebra/native/sw_bls12377/g1.go:609]: Yes indeed we can follow the pattern in `MultiScalarMul`

File: std/algebra/native/sw_bls12377/g1.go [hunk 2]
```diff
@@ -453,33 +511,157 @@ func (p *G1Affine) DoubleAndAdd(api frontend.API, p1, p2 *G1Affine) *G1Affine {
 
 // ScalarMulBase computes s * g1 and returns it, where g1 is the fixed generator. It doesn't modify s.
 func (P *G1Affine) ScalarMulBase(api frontend.API, s frontend.Variable) *G1Affine {
+	_, _, g1aff, _ := bls12377.Generators()
```
[the remark this answers is not in the document]
Comment (yelhousni) [reply std/algebra/native/sw_bls12377/g1.go:514]: In my benchmarks, for the native case, using the endomorphism (`varScalarMul`) resulted in less constraints than using precomputed precomputed points with a bigger loop (128 vs. 253). For the emulated case it's the opposite as emulated arithmetic is costly. As a small optimization for the native case, we can pre-compute `± Q ± Φ(Q)`.

File: std/algebra/native/sw_bls12377/pairing.go [hunk 3]
```diff
@@ -29,35 +29,45 @@ type GT = fields_bls12377.E12
...
 // MillerLoop computes the product of n miller loops (n can be 1)
 // ∏ᵢ { fᵢ_{x₀,Q}(P) }
 func MillerLoop(api frontend.API, P []G1Affine, Q []G2Affine) (GT, error) {
+
 	// check input size match
 	n := len(P)
 	if n == 0 || n != len(Q) {
 		return GT{}, errors.New("invalid inputs sizes")
 	}
+	lines := make([]lineEvaluations, len(Q))
```
[the remark this answers is not in the document]
Comment (yelhousni) [reply std/algebra/native/sw_bls12377/pairing.go:41]: Makes sense. We can change `MillerLoop` to take `[]*G2Affine` instead.

File: std/commitments/kzg/verifier.go [hunk 4]
```diff
@@ -702,10 +683,11 @@ func (v *Verifier[FR, G1El, G2El, GTEl]) fold(digests []Commitment[G1El], fai []
 	for i := range digestPoints {
 		digestPoints[i] = &digests[i].G1El
 	}
-	foldedDigest, err := v.curve.MultiScalarMul(digestPoints, ci)
+	foldedDigest, err := v.curve.MultiScalarMul(digestPoints[1:], ci[1:])
 	if err != nil {
 		return Commitment[G1El]{}, nil, fmt.Errorf("fold digests: %w", err)
 	}
+	foldedDigest = v.curve.Add(foldedDigest, digestPoints[0])
```
[the remark this answers is not in the document]
Comment (yelhousni) [reply std/commitments/kzg/verifier.go:690]: Agree. Here I make the assumption that `fold` method is only used in `foldedDigests, foldedEvals, err := v.fold(digests, evals, randomNumbers)` where `randomNumbers[0] = v.scalarApi.One()`. We need to add a doc comment, but should keep this optimization as it is 1 (emulated) point add vs. ~380 (emulated) point add.

File: std/commitments/kzg/verifier.go [hunk 5]
```diff
@@ -632,8 +613,8 @@ func (v *Verifier[FR, G1El, G2El, GTEl]) FoldProof(digests []Commitment[G1El], b
 	for i := 2; i < nbDigests; i++ {
 		gammai[i] = v.scalarApi.Mul(gammai[i-1], gamma)
 	}
-	foldedEvaluations := v.scalarApi.Zero()
-	for i := 0; i < nbDigests; i++ {
+	foldedEvaluations := &batchOpeningProof.ClaimedValues[0]
```
[the remark this answers is not in the document]
Comment (yelhousni) [reply std/commitments/kzg/verifier.go:616]: Agree

**Comment by yelhousni** [comment]
> The PR is good to merge. We can address my other issues in other PR. I think this PR needs to be merged to unblock #960.

Okay let's merge this. Do you want to merge https://github.com/Consensys/gnark/pull/874 first and then this or just this and close 874?

**Comment by ivokub** [comment]
> > The PR is good to merge. We can address my other issues in other PR. I think this PR needs to be merged to unblock #960.
> 
> Okay let's merge this. Do you want to merge https://github.com/Consensys/gnark/pull/874 first and then this or just this and close 874?
> 

It seems all changes are already incorporated here. I'd merge only this and then close 874.

File: std/algebra/native/sw_bls12377/g1.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (yelhousni) [reply std/algebra/native/sw_bls12377/g1.go:514]: - 3858 scs for `varScalarMul` and 4031 scs for `ScalarMulBase` with precomputed doubles.

[the remark this answers is not in the document]
Comment (yelhousni) [reply std/algebra/native/sw_bls12377/g1.go:514]: - 2-chain case: 
3,858 scs for `varScalarMul` and 4031 scs for `ScalarMulBase` with precomputed doubles.
- Emulated BW6 in BN254 case: 
2,291,619 scs for `varScalarMul` and 1,882,197 scs for `ScalarMulBase` with precomputed doubles.

[the remark this answers is not in the document]
Comment (yelhousni) [reply std/algebra/native/sw_bls12377/g1.go:514]: - 2-chain case: 
3,858 scs for `varScalarMul` and 4031 scs for `ScalarMulBase` with precomputed doubles.
- Emulated BW6 in BN254 case: 
2,291,619 scs for `varScalarMul` and 1,882,197 scs for `ScalarMulBase` with precomputed doubles.

**Pull request merged**